core/src/test/java/org/infinispan/functional/decorators/FunctionalAdvancedCache.java

```diff
@@ -643,4 +645,186 @@ public void removeListener(Object listener) {
       }
    }
 
+   private static final class SetAsCacheSet<E> implements CacheSet<E> {
+      final Set<E> set;
+
+      private SetAsCacheSet(Set<E> set) {
+         this.set = set;
+      }
+
+      @Override
+      public CacheStream<E> stream() {
+         return null;
+      }
+
+      @Override
+      public CacheStream<E> parallelStream() {
+         return null;
+      }
+
+      @Override
+      public int size() {
+         return set.size();
+      }
+
+      @Override
+      public boolean isEmpty() {
+         return set.isEmpty();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+         return set.contains(o);
+      }
+
+      @Override
+      public CloseableIterator<E> iterator() {
+         return Closeables.iterator(set.iterator());
+      }
+
+      @Override
+      public Object[] toArray() {
+         return set.toArray();
+      }
+
+      @Override
+      public <T> T[] toArray(T[] a) {
+         return set.toArray(a);
+      }
+
+      @Override
+      public boolean add(E e) {
+         return set.add(e);
+      }
+
+      @Override
+      public boolean remove(Object o) {
+         return set.remove(o);
+      }
+
+      @Override
+      public boolean containsAll(Collection<?> c) {
+         return set.containsAll(c);
+      }
+
+      @Override
+      public boolean addAll(Collection<? extends E> c) {
+         return set.addAll(c);
+      }
+
+      @Override
+      public boolean removeAll(Collection<?> c) {
+         return set.removeAll(c);
+      }
+
+      @Override
+      public boolean retainAll(Collection<?> c) {
+         return set.retainAll(c);
+      }
+
+      @Override
+      public void clear() {
+         set.clear();
+      }
+
+      @Override
+      public CloseableSpliterator<E> spliterator() {
+         return null;
+      }
+
+      @Override
+      public String toString() {
+         return "SetAsCacheSet{" +
+            "set=" + set +
+            '}';
+      }
+   }
+
+   private static class CollectionAsCacheCollection<E> implements CacheCollection<E> {
```

Comment:
You should be able to use `AbstractDelegatingCollection` as the base class here and then only have to implement the CacheCollection specific override methods.